app/grandchallenge/components/tasks.py

```diff
@@ -1139,3 +1139,106 @@ def validate_voxel_values(*, civ_pk):
             civ.image.save()
 
     civ.interface._validate_voxel_values(civ.image)
+
+
+@shared_task(
+    **settings.CELERY_TASK_DECORATOR_KWARGS["acks-late-micro-short"],
+)
+def add_image_to_object(
+    *,
+    app_label,
+    model_name,
+    object_pk,
+    interface_pk,
+    upload_session_pk,
+):
+    from grandchallenge.components.models import (
+        ComponentInterface,
+        ComponentInterfaceValue,
+    )
+
+    with transaction.atomic():
+        object = get_model_instance(
+            pk=object_pk,
+            app_label=app_label,
+            model_name=model_name,
+        )
+        interface = ComponentInterface.objects.get(pk=interface_pk)
+        upload_session = RawImageUploadSession.objects.get(
+            pk=upload_session_pk
+        )
+        try:
+            image = Image.objects.get(origin_id=upload_session_pk)
+        except (Image.DoesNotExist, Image.MultipleObjectsReturned):
+            error_message = "Image imports should result in a single image"
+            upload_session.status = RawImageUploadSession.FAILURE
+            upload_session.error_message = error_message
+            upload_session.save()
+            return
+
+        object.values.remove(*object.values.filter(interface=interface))
+        civ, created = ComponentInterfaceValue.objects.get_or_create(
+            interface=interface, image=image
+        )
+        if created:
+            try:
+                civ.full_clean()
+            except ValidationError as e:
+                # this should only happen for new uploads
+                upload_session.status = RawImageUploadSession.FAILURE
+                upload_session.error_message = e.message
+                upload_session.save()
+        object.values.add(civ)
+
+
+@shared_task(**settings.CELERY_TASK_DECORATOR_KWARGS["acks-late-2xlarge"])
+def add_file_to_object(
+    *,
+    app_label,
+    model_name,
+    user_upload_pk,
+    object_pk,
+    interface_pk,
+    civ_pk=None,
+):
+    from grandchallenge.components.models import (
+        ComponentInterface,
+        ComponentInterfaceValue,
+    )
+
+    user_upload = UserUpload.objects.get(pk=user_upload_pk)
+    object = get_model_instance(
+        pk=object_pk,
+        app_label=app_label,
+        model_name=model_name,
+    )
+    interface = ComponentInterface.objects.get(pk=interface_pk)
+    error = None
+    with transaction.atomic():
+        civ = ComponentInterfaceValue.objects.create(interface=interface)
```

Comment:
Why use create and then rollback on error? Better to instantiate, clean and handle error/save, no?

Reply:
That indeed makes more sense. Again a copy-paste oversight.